esp-hal/src/spi/master.rs

```diff
@@ -618,9 +623,24 @@ where
 
     /// Assign the MISO (Master In Slave Out) pin for the SPI instance.
     ///
+    /// Enables input functionality for the pin, and connects it to the MISO
+    /// signal.
+    pub fn with_miso<MISO: PeripheralInput>(self, miso: impl Peripheral<P = MISO> + 'd) -> Self {
+        crate::into_mapped_ref!(miso);
+        miso.enable_input(true, private::Internal);
+
+        self.driver().miso.connect_to(&mut miso);
+
+        self
+    }
+
+    /// Assign the SIO1/MISO pin for the SPI instance.
+    ///
     /// Enables both input and output functionality for the pin, and connects it
     /// to the MISO signal and SIO1 input signal.
-    pub fn with_miso<MISO: PeripheralOutput>(self, miso: impl Peripheral<P = MISO> + 'd) -> Self {
+    ///
+    /// Note: You do not need to call [Self::with_miso] when this is used.
+    pub fn with_sio1<SIO1: PeripheralOutput>(self, miso: impl Peripheral<P = SIO1> + 'd) -> Self {
```

Comment:
> * None of our documentation (I mean, outside of esp-hal) actually refers to the signal as SIO1, I think.

From the S3 TRM, Table 30-3 "FSPIQ | SPI3_Q | MISO/SIO1 (serial data input and output, bit1)".
Though it's the only mention of it.

> MISO being singled out may be somewhat inconsistent.

The inconsistency is coming from the hardware, not the hal.

> Depending on whether we want to support splitting signals between pins

Sending a peripheral's output signal to multiple pins is much more useful and will probably look very similar. Might as well support both before cementing the API :slightly_smiling_face:

Reply:
> Sending a peripheral's output signal to multiple pins is much more useful

It is possible to connect different output signals, `OutputSignal::connect_to` does just that. As the Instance traits are public and they contain the peripheral signals, I don't think we need to do really anything to enable that case. We don't really bind the pin to the signal, so I guess from a type safety standpoint we might change that, but the functionality itself is there. As this is probably not a common use case, I'd not change the peripheral drivers more, except document the possibility.

I am not quite sure about naming. Ideally I'd like to avoid multiple functions that have different names but do the exact same thing. But I have to acknowledge that QSPI/OSPI naming differs from SPI, and that the same bus can be used to connect different devices, so we need to cover all this in a single driver.

I guess considering this, `with_sio1` makes sense. I really, really resist the idea of `with_sio0`, though, but it just looks wrong not to have it if the user really wants to configure a QSPI/Octal SPI bus.

Well, there's my long rant that includes me coming around to this idea.

Reply:
Regarding the multiple pins stuff. I had imagined something along the lines of `with_miso(OutputGroup([peripherals.GPIO1, peripherals.GPIO2, ....]))`. So the drivers won't change, it'd be just a new implementation of `PeripheralOutput`.

> It is possible to connect different output signals, `OutputSignal::connect_to` does just that.

Interesting, I didn't realize that the intention was for these to be public. It has some interesting consequences, ownership wise. I'll open a separate issue about this.

Reply:
>I had imagined something along the lines of with_miso(OutputGroup([peripherals.GPIO1, peripherals.GPIO2, ....])). So the drivers won't change, it'd be just a new implementation of PeripheralOutput.

I've thought about this but since pins are different types, it's not this simple. Tuples would work,

Reply:
To me
```rust
.with_mosi(p1)
.with_sio1(p2)
```
looks weird.

But I have to admit also this
```rust
.with_mosi(p1)
.with_miso(p2)
.with_sio2(p3)
.with_sio3(p4)
```

already looked weird - so it's not worse at least